src/builders/utils/unimodules.ts

```diff
@@ -0,0 +1,162 @@
+import fs from 'fs-extra';
+import filter from 'lodash/filter';
+import keys from 'lodash/keys';
+import merge from 'lodash/merge';
+import values from 'lodash/values';
+import path from 'path';
+import logger from 'turtle/logger';
+
+export interface IUnimoduleEntry {
+  name: string;
+  version: string;
+  dirname: string;
+}
+
+interface IPackageInfo {
+  name: string;
+  version: string;
+  dirname: string; // can be difrent than name
+  dependencies: string[];
+}
+
+export async function resolveNativeModules(workingdir: string, packages?: string[]): Promise<IUnimoduleEntry[]> {
+  const resolver = new Resolver(workingdir);
+  await resolver.init();
+
+  if (!packages) {
+    resolver.addAll();
+  } else {
+    const corePackages = await readCorePackages(workingdir);
+    for (const pkg of corePackages) {
+      if (resolver.isUnimodule(pkg)) {
+        resolver.addModule(pkg);
+      }
+    }
+    for (const pkg of packages) {
+      if (resolver.isUnimodule(pkg)) {
+        resolver.addModule(pkg);
+      }
+    }
+  }
+
+  return resolver.getModules();
+}
+
+async function readCorePackages(workingdir: string) {
+  const pkgJsonPath = path.join(workingdir, 'node_modules', 'react-native-unimodules', 'package.json');
+  if (await fs.pathExists(pkgJsonPath)) {
+    const packageJson = JSON.parse(await fs.readFile(pkgJsonPath, 'utf-8'));
+    return keys(packageJson.dependencies);
+  }
+  logger.warn('No core packages detected');
+  return [];
+}
+
+class Resolver {
+  private workingdir: string;
+  private modulesMap: {[key: string]: IUnimoduleEntry};
+  private dependencyMap: {[key: string]: IPackageInfo};
+
+  constructor(workingdir: string) {
+    this.workingdir = workingdir;
+    this.modulesMap = {};
+    this.dependencyMap = {};
+  }
+
+  public async init(): Promise<void> {
+    this.dependencyMap = await generateDependencyMap(this.workingdir);
+  }
+
+  public isUnimodule(moduleName: string): boolean {
+    return !!this.dependencyMap[moduleName];
+  }
+
+  public addAll() {
+    for (const moduleName of Object.keys(this.dependencyMap)) {
+      this.addModule(moduleName);
+    }
+  }
+
+  public addModule(moduleName: string) {
+    if (this.dependencyMap[moduleName]) {
+      this.resolveModule(moduleName);
+    } else {
+      throw new Error('Attempting to add unknown module');
+    }
+  }
+
+  public getModules(): IUnimoduleEntry[] {
+     return values(this.modulesMap);
+  }
+
+  private async resolveModule(moduleName: string) {
+    if (this.modulesMap[moduleName]) {
+      return;
+    }
+    const pkg = this.dependencyMap[moduleName];
+    logger.info(`Adding ${moduleName}:${pkg.version}`);
+    this.modulesMap[moduleName] = {
+      name: pkg.name,
+      dirname: pkg.dirname,
+      version: pkg.version,
+    };
+    for (const dep of pkg.dependencies) {
+      this.resolveModule(dep);
+    }
+  }
+}
+
+async function generateDependencyMap(workingdir: string): Promise<{[key: string]: IPackageInfo}> {
```

Comment:
```suggestion
async function generateDependenciesInfo(directory: string): Promise<{[key: string]: IPackageInfo}> {
```
Maybe it's only my taste, but I would make this method more generic, i. e. it would return dependencies info only for one directory and merge the results in `Resolver` initializer, which would accept a list of possible locations of unimodules

Reply:
This is very internal logic, so changing this in the future will be very easy and it won't introduce any breaking changes. I think it's too early to do this now.